compiler/qsc_formatter/src/formatter.rs

```diff
@@ -121,36 +133,124 @@ fn apply_rules(
     left: &ConcreteToken,
     whitespace: &str,
     right: &ConcreteToken,
-    code: &str,
-    indent_level: usize,
+    state: &mut FormatterState,
 ) -> Vec<TextEdit> {
     let mut edits = vec![];
     // when we get here, neither left nor right should be whitespace
 
-    // if the right is a close brace, the indent level should be one less
-    let indent_level = if let ConcreteTokenKind::Syntax(TokenKind::Close(Delim::Brace)) = right.kind
-    {
-        indent_level.saturating_sub(1)
-    } else {
-        indent_level
-    };
-
-    let new_line_in_spaces = whitespace.contains('\n');
+    let are_newlines_in_spaces = whitespace.contains('\n');
+    let mut newline_context = state
+        .delim_newlines_stack
+        .last()
+        .map_or(NewlineContext::NoContext, |b| *b);
 
     use qsc_frontend::keyword::Keyword;
     use qsc_frontend::lex::cooked::ClosedBinOp;
     use ConcreteTokenKind::*;
     use TokenKind::*;
+
```

Comment:
With the two new `match` expressions, this function is quite long - would you extracting out a couple of helper functions? `update_type_param_state` and `update_delim_state` maybe

Reply:
Done